generator/src/main/java/com/scottlogic/deg/generator/decisiontree/ConstraintNodeBuilder.java

```diff
@@ -0,0 +1,107 @@
+package com.scottlogic.deg.generator.decisiontree;
+
+import com.scottlogic.deg.common.profile.constraints.atomic.AtomicConstraint;
+import com.scottlogic.deg.common.util.FlatMappingSpliterator;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static java.util.stream.Stream.concat;
+
+public class ConstraintNodeBuilder {
+    private final Collection<AtomicConstraint> atomicConstraints;
+    private final Collection<DecisionNode> decisions;
+    private final Set<NodeMarking> nodeMarkings;
+
+    public ConstraintNodeBuilder(Collection<AtomicConstraint> atomicConstraints, Collection<DecisionNode> decisions, Set<NodeMarking> nodeMarkings) {
+        this.atomicConstraints = atomicConstraints;
+        this.decisions = decisions;
+        this.nodeMarkings = nodeMarkings;
+    }
+
+    public ConstraintNodeBuilder() {
+        this(Collections.emptyList(), Collections.emptyList(), Collections.emptySet());
+    }
+
+    public ConstraintNodeBuilder setAtomicConstraints(Collection<AtomicConstraint> newAtomicConstraints) {
```

Comment:
I might be wrong, but aren't builder function names normally of the form `with...`, e.g. `withAtomicConstraints`, `withoutAtomicConstraint`, `withNodeMarking` and `withNodeMarkings`?

Reply:
the issue being, it currently also has "addX" methods.

addX and withX aren't distinct enough.

addX and setX are distinct.

it shouldn't really have the addX methods, but that's a larger refactor effort.